### PR TITLE
Add support for retrieving contract ABIs directly from Etherscan

### LINKED ISSUE #119: Generate Contract Bindings for `etherscan.io` Verified Contracts.

### PR DESCRIPTION
**Pull request by nlordell**
This PR adds initial support for retrieving contract ABIs directly from Etherscan using the HTTP API.

Closes #119 

### Test Plan

A couple unit tests and a new example.

### CODE REVIEW

File: derive/src/lib.rs [hunk 1]
```diff
@@ -24,10 +25,22 @@ use syn::{braced, parse_macro_input, Error as SynError, Ident, LitInt, LitStr, T
 /// ethcontract::contract!("build/contracts/MyContract.json");
 /// ```
 ///
+/// Alternatively, an etherscan URL can be specified. In this case the ABI will
+/// be retrieved and used to generate type-safe contract bindings:
+///
+/// ```ignore
+/// ethcontract::contract!("etherscan:0x0001020304050607080910111213141516171819");
+/// // or
+/// ethcontract::contract!("https://etherscan.io/address/0x0001020304050607080910111213141516171819");
```
Comment (fleupold) [inline derive/src/lib.rs:34]: Would this work for rinkeby etherscan as well?

**Review by fleupold — COMMENTED** [review]
I wonder if we even have to leak the fact that we are getting this from etherscan? Maybe the API could be to just provide the network and address and then the library tries to fetch the data (by looking it up on etherscan or other source registries)

**Review by e00E — COMMENTED** [review]
Code looks good.

Its questionable whether any build should rely on an external http service at compile time. It could be better to have the download of the abi from ethcontract as its own functionality and then use the abi json file on disk as usual.
This was probably talked about when making the issue but I didn't see any discussion on Github so I wanted to throw this thought out there.

File: derive/src/lib.rs (hunk 1 above)
Comment (nlordell) [reply derive/src/lib.rs:34]: Currently no, but we can add support for it. I would probably want to do it in a separate PR though.

**Comment by nlordell** [comment]
> Its questionable whether any build should rely on an external http service at compile time.

I think the argument here is that you are using it as a package manager (which is also an HTTP service). And since Etherscan and the blockchain are immutable, then I feel like it is OK to use it in that way.

I think if you wanted to use ABIs from Etherscan and vendor them, then you would just include them as JSON files which we already support.

**Comment by nlordell** [comment]
> I wonder if we even have to leak the fact that we are getting this from etherscan? Maybe the API could be to just provide the network and address and then the library tries to fetch the data (by looking it up on etherscan or other source registries)

I wanted to make it very clear that this is reaching out to an external service to get this information and not do it sneakily against people's expectations. Also, I plan on adding ethPM and npm support next so I thought that using URLs as I have done was a good way to make it unambiguous.

**Review by fleupold — APPROVED** [review]
looks good...

/magic

**Review by e00E — APPROVED** [review]

**Review by bh2smith — APPROVED** [review]
I guess this would require that the contract has been verified on etherscan. Very cool. Watch out truffle, here comes ethcontracts-rs!

**Pull request merged**